[Serve] Add the `SERVE_REQUEST_PROCESSING_TIMEOUT_S` environment variable

shrekris-anyscale opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

Under some circumstances (e.g. see [Kuberay issue #615](https://github.com/ray-project/kuberay/issues/615)), the `HTTPProxy` actor tends to hang while waiting for requests to be processed. This change introduces the `SERVE_REQUEST_PROCESSING_TIMEOUT_S` environment variable, so users can control how long their `HTTPProxies` should wait for a request to finish before retrying at another replica.

As an example, when `SERVE_REQUEST_PROCESSING_TIMEOUT_S` is set to 1.5 seconds, the query latency goes down from ~20 seconds (the time it takes for the GCS to recover) to ~1.5 seconds (the time it takes to retry the query). See the log outputs below for more details. In general, without this env var, requests hang until the GCS recovers. The env var limits the hanging overhead by whatever value it's set to.

#### Sample log outputs

These results were produced with the following setup on KubeRay:

* 1 head pod: 1 HTTPProxy + 1 deployment replica + 1 Serve controller
* 1 worker pod: 1 HTTPProxy + 1 deployment replica

The head pod was killed manually, and while it was recovering, requests were sent via port-forward directly to the worker pod. These are the HTTPProxy's logs while processing those requests.

Without the env var:

```console
...
DEBUG 2022-10-13 14:12:15,017 http_proxy 10.68.2.112 router.py:146 - Assigned query qoEjwkdjSc to replica Loud#EsXyss.
DEBUG 2022-10-13 14:12:35,069 http_proxy 10.68.2.112 router.py:146 - Assigned query SRnxTLejNu to replica Loud#sVOznE.
INFO 2022-10-13 14:12:35,102 http_proxy 10.68.2.112 http_proxy.py:320 - GET / 200 20057.4ms
...
```

Replica `EsXyss` got killed. Replica `sVOznE` stayed alive. There’s ~20 seconds before the HTTP proxy retried when it sent a request to `EsXyss`.

With the env var set to 1.5:

```console
...
DEBUG 2022-11-03 18:26:42,035 http_proxy 10.68.1.24 router.py:145 - Assigned query aHAATPlqtn to replica Loud#kKRCAK.
DEBUG 2022-11-03 18:26:43,537 http_proxy 10.68.1.24 http_proxy.py:91 - Request didn't finish within 1.5 seconds. Retrying with another replica. You can modify this timeout by setting the "REQUEST_PROCESSING_TIMEOUT_S" env var.
DEBUG 2022-11-03 18:26:43,538 http_proxy 10.68.1.24 router.py:145 - Assigned query tfhQwDYLhU to replica Loud#mpgmZx.
INFO 2022-11-03 18:26:43,543 http_proxy 10.68.1.24 http_proxy.py:358 - GET / 200 1508.3ms
...
```

Replica `aHAATPlqtn` got killed. Replica `mpgmZx` stayed alive. There's ~1.5 seconds before the HTTP proxy retried when it sent a request to `mpgmZx`.

## Related issue number

<!-- For example: "Closes #1234" -->

Addresses [Kuberay issue #615](https://github.com/ray-project/kuberay/issues/615).

## Checks

- [X] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
     - This change adds the `TestRequestProcessingTimeoutS` test class.

simon-mo commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -36,6 +36,9 @@
     os.environ.get("SERVE_SOCKET_REUSE_PORT_ENABLED", "1") == "1"
 )
+PROXY_FORWARD_ATTEMPT_TIMEOUT_S = (
+    float(os.environ.get("PROXY_FORWARD_ATTEMPT_TIMEOUT_S", 0)) or None
```
```suggestion
    float(os.environ.get("SERVE_PROXY_FORWARD_ATTEMPT_TIMEOUT_S", 0)) or None
```

shrekris-anyscale replied:
Thanks, I made the change.

simon-mo reviewed (changes requested):
This is just timeing out when picking a replica (i.e. the assignment task). can you verify this actually fixes the issue? There are two more awaits down there that's actually interacting with the replicas
```
        try:
            object_ref = await assignment_task
            result = await object_ref
            client_disconnection_task.cancel()
            break
```

shrekris-anyscale commented:
> This is just timeing out when picking a replica (i.e. the assignment task). can you verify this actually fixes the issue? There are two more awaits down there that's actually interacting with the replicas

@simon-mo Good call– I tried manually verifying if the fix works on KubeRay, but it turns out the hang was indeed in one of the other `await` statements. I updated the change, so the timeout happens while the proxy is waiting for the request to be processed, not only while the proxy is assigning the request.

I manually verified that this fix works on KubeRay, and I added a unit test to make sure that this doesn't impede normal operation.

sihanwang41 commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -75,9 +78,21 @@ async def _send_request_to_handle(handle, scope, receive, send) -> str:
-            result = await object_ref
-            client_disconnection_task.cancel()
-            break
+            _, timed_out = await asyncio.wait(
```
cancel time out tasks.

shrekris-anyscale replied:
I canceled the `client_disconnect_task`. I'm not sure how to cancel the `object_ref` from `timed_out`. The `object_ref` is [a Ray object ref](https://github.com/ray-project/ray/blob/437cf14e737df679ccf71a0e38d52f37b0e436fa/python/ray/serve/_private/router.py#L174-L178), not an `asyncio.Task`.

I believe the only ways to terminate a Ray remote call are [`ray.kill`](https://docs.ray.io/en/releases-2.0.1/ray-core/package-ref.html#ray-kill) (which would kill the whole replica) or [`ray.cancel`](https://docs.ray.io/en/releases-2.0.1/ray-core/package-ref.html#ray-cancel) (which can't be used on Ray actor method calls). I don't think there's a straightforward way to kill the call. Do you have any suggestions on how to kill it?

sihanwang41 replied:
good call, i don't think we are able to do it. We can leave it as it is.

sihanwang41 commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -75,9 +78,21 @@ async def _send_request_to_handle(handle, scope, receive, send) -> str:
+            _, timed_out = await asyncio.wait(
+                [object_ref], timeout=REQUEST_PROCESSING_TIMEOUT_S
+            )
+            if timed_out:
```
the logic here is ignoring `backoff_time_s`?

shrekris-anyscale replied:
Thanks for the catch, I added the `backoff_time_s`.

simon-mo commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -36,6 +36,9 @@
     os.environ.get("SERVE_SOCKET_REUSE_PORT_ENABLED", "1") == "1"
 )
+REQUEST_PROCESSING_TIMEOUT_S = (
+    float(os.environ.get("REQUEST_PROCESSING_TIMEOUT_S", 0)) or None
```
```suggestion
    float(os.environ.get("SERVE_REQUEST_PROCESSING_TIMEOUT_S", 0)) or None
```

shrekris-anyscale replied:
I changed `REQUEST_PROCESSING_TIMEOUT_S` to `SERVE_REQUEST_PROCESSING_TIMEOUT_S`.

simon-mo commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -75,9 +78,21 @@ async def _send_request_to_handle(handle, scope, receive, send) -> str:
+                [object_ref], timeout=REQUEST_PROCESSING_TIMEOUT_S
+            )
+            if timed_out:
+                logger.debug(
```
info is fine here because it is rare

shrekris-anyscale replied:
Sounds good, I upgraded it to `info`.

simon-mo commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -75,9 +78,21 @@ async def _send_request_to_handle(handle, scope, receive, send) -> str:
+                logger.debug(
+                    f"Request didn't finish within {REQUEST_PROCESSING_TIMEOUT_S} "
+                    "seconds. Retrying with another replica. You can modify "
+                    'this timeout by setting the "REQUEST_PROCESSING_TIMEOUT_S" '
```
```suggestion
                    'this timeout by setting the "SERVE_REQUEST_PROCESSING_TIMEOUT_S" '
```

shrekris-anyscale replied:
I updated the name.

simon-mo commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -75,9 +78,21 @@ async def _send_request_to_handle(handle, scope, receive, send) -> str:
+                    'this timeout by setting the "REQUEST_PROCESSING_TIMEOUT_S" '
+                    "env var."
+                )
+                retries += 1
```
no backoff? the other else clause has `backoff_time_s *= 1.5`

simon-mo replied:
also missing `client_disconnection_task.cancel()`

shrekris-anyscale replied:
Thanks for the catches, I added the `backoff_time_s` and the `client_disconnection_task.cancel()`.

simon-mo reviewed (commented):
please leave a note saying how can the release test be improved to catch this (while you still have the context), we don't have to update it now, but would be useful for the future

shrekris-anyscale commented:
Thanks for reviewing the change @simon-mo and @sihanwang41– I've addressed your comments. @simon-mo I've added a note explaining the edge case solved by the new env var and how to test it in a release test. Please let me know if you have any more feedback.

sihanwang41 commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -75,9 +78,32 @@ async def _send_request_to_handle(handle, scope, receive, send) -> str:
+                    'setting the "SERVE_REQUEST_PROCESSING_TIMEOUT_S" env var.'
+                )
+                client_disconnection_task.cancel()
+                backoff_time_s *= 1.5
```
we only add backoff_time_s but never sleep

shrekris-anyscale replied:
Good catch, I moved all the backoff logic to a single conditional to avoid these omissions.

sihanwang41 commented on python/ray/serve/_private/http_proxy.py:
```diff
@@ -75,9 +78,32 @@ async def _send_request_to_handle(handle, scope, receive, send) -> str:
+                    'setting the "SERVE_REQUEST_PROCESSING_TIMEOUT_S" env var.'
+                )
+                client_disconnection_task.cancel()
+                backoff_time_s *= 1.5
```
There are some duplicate codes here and `RayActorError`, we can raise RayServeException here and do all the backoff & sleep logics inside the exception?

RayServeException and RayActorError can use same code.

shrekris-anyscale replied:
Good idea, I moved all the backoff logic to a single conditional.

shrekris-anyscale commented:
The failing Windows tests ([`test_runtime_env_working_dir`](https://buildkite.com/ray-project/oss-ci-build-pr/builds/4434#01844fc8-20cf-46b8-8afb-3fa4957948bb/6343-8239) and [`test_cancel`](https://buildkite.com/ray-project/oss-ci-build-pr/builds/4434#01844fc8-20d1-4421-b0d2-819d4d0fdcd2/9877-11664)) are unrelated.
